specification/dartLangSpec.tex

```diff
@@ -16231,238 +16306,445 @@ \subsection{Imports}
+and \NamespaceName{\metavar{prefix},i} is the namespace obtained by
+adding a binding of the name \code{loadLibrary}
+to an implicitly induced declaration of a function with signature
+\code{Future<\VOID> loadLibrary()}
+to \NamespaceName{\metavar{exported},1}.
 
-where $hide(l, n)$ takes a list of identifiers $l$ and a namespace $n$, and produces a namespace that is identical to $n$ except that for each name $k$ in $l$, $k$ and \code{$k$=} are undefined.
-\end{itemize}
+\LMHash{}%
+Then \NamespaceName{\metavar{one},i} is a namespace that has
+a single binding that maps $p_i$ to \NamespaceName{\metavar{prefix},i}.
 
 \LMHash{}%
-If $I$ includes a prefix clause of the form \AS{} $p$,
-let $NS = \{p: prefixObject(NS_n)\}$,
-where $prefixObject(NS_n)$ is a \Index{prefix object} for the namespace $NS_n$,
-which is an object that has the following members:
+In this situation we say that \NamespaceName{\metavar{one},i} is a
+\Index{prefix namespace},
+because it maps a library prefix to a namespace.
 
-\begin{itemize}
-\item For every top level function $f$ named \id{} in $NS_n$, a corresponding method with the same name and signature as $f$ that forwards (\ref{functionDeclarations}) to $f$.
-\item For every top level getter with the same name and signature as $g$ named \id{} in $NS_n$, a corresponding getter that forwards to $g$.
-\item For every top level setter $s$ with the same name and signature as named \id{} in $NS_n$, a corresponding setter that forwards to $s$.
-\item For every type $T$ named \id{} in $NS_n$, a corresponding getter named \id{} with return type \code{Type}, that, when invoked, returns the type object for $T$.
-\end{itemize}
+\commentary{%
+A prefix namespace is not a Dart object,
+it is merely a device which is used to manage
+expressions of the form \synt{qualifiedName} and what they refer to
+during static analysis.
+Consequently, any attempt to use a prefix namespace as an object
+is a compile-time error, e.g.,
+it cannot be the result of an expression evaluation.
+
+Note that if $l$ and $q$ are such that
+$I_l$ and $I_q$ both have the same prefix $p$
+then
+$\NamespaceName{\metavar{one},l} = \NamespaceName{\metavar{one},q} =
+\NamespaceName{\metavar{one},l} \cup \NamespaceName{\metavar{one},q}$.%
+}
+\EndCase
 
 \LMHash{}%
-It is a compile-time error if the current library declares a top-level member named $p$.
+{\bf Step two.}
+In the second step we resolve top level conflicts
+among the namespaces obtained in the first step.
 
-The static type of the prefix object $p$ is a unique interface type that has those members whose names and signatures are listed above.
-% What is the static type of a prefix object. Need to state that is a(n anonymous) type that has members with the same names as signatures as above.
+\LMHash{}%
+Let \NamespaceName{\metavar{merged}} be the result of applying
+conflict merging to the namespaces
+\List{\metavar{NS}}{\metavar{one},1}{\metavar{one},m}.
```

Comment:
> These namespaces contain single-entry namespaces, where
> entry's value is a namespace

Yes, but in every case where that's true for a name `n` there is no conflict with any other namespaces where `n` is a declaration, because they have been removed in step one. And there's no conflict with any other namespace where `n` is a prefix namespace, because `n` maps to the exact same namespace for all imports with the same prefix.

> Conflict merging is only well-defined for namespaces with
> values that are declarations

Plus values that are prefix namespaces, as long as they agree.

> Also, this would cause an import prefix to conflict with an imported name,
> which is wrong

No, the imported name was removed in step one, so there is no conflict.

> The prefix is a local declaration and should shadow the import

We can't treat the prefix exactly like a local declaration because there may be multiple imports with the same prefix. But we do remove the imported names that are in conflict with import prefixes in step one.

So conflict merging is not applied to arguments where it is ill defined, and the resulting namespace does include both the names imported with a prefix and all the prefix namespaces.